File: rs/src/gameboy.rs
use anyhow::Result;

use crate::apu;
use crate::args;
use crate::buttons;
use crate::cart;
use crate::clock;
use crate::cpu;
use crate::gpu;
use crate::ram;

pub struct GameBoy<'a> {
    ram: ram::RAM,
    cpu: cpu::CPU,
    gpu: gpu::GPU<'a>,
    apu: apu::APU,
    buttons: buttons::Buttons,
    clock: clock::Clock,
}

impl<'a> GameBoy<'a> {
    #[inline(never)]
    pub fn new(args: args::Args) -> Result<GameBoy<'a>> {
        let sdl = sdl2::init().map_err(anyhow::Error::msg)?;

        let cart = cart::Cart::new(args.rom.as_str())?;
        let cart_name = cart.name.clone();
        let ram = ram::RAM::new(cart)?;
        let cpu = cpu::CPU::new(args.debug_cpu);
        let gpu = gpu::GPU::new(&sdl, cart_name, args.headless, args.debug_gpu)?;
        let apu = apu::APU::new(&sdl, args.silent, args.debug_apu)?;
        let buttons = buttons::Buttons::new(sdl, args.headless)?;
        let clock = clock::Clock::new(args.profile, args.turbo);

        Ok(GameBoy {
            ram,
            cpu,
            gpu,
            apu,
            buttons,
            clock,
        })
    }

    #[inline(never)]
    pub fn run(&mut self) -> Result<()> {
        loop {
            self.tick()?;
        }
    }

    #[inline(always)]
    pub fn tick(&mut self) -> Result<()> {
        self.cpu.tick(&mut self.ram)?;
        self.gpu.tick(&mut self.ram, &mut self.cpu)?;
        self.buttons.tick(&mut self.ram, &mut self.cpu)?;
        self.clock.tick(&self.buttons)?;
        self.apu.tick(&mut self.ram);

        Ok(())
    }
}

File: c/src/gameboy.c
#include "gameboy.h"

void gameboy_ctor(struct GameBoy *self, struct Args *args) {
    cart_ctor(&self->cart, args->rom, false);
    ram_ctor(&self->ram, &self->cart, args->debug_ram);
    cpu_ctor(&self->cpu, &self->ram, args->debug_cpu);
    gpu_ctor(&self->gpu, &self->cpu, &self->ram, self->cart.name, args->headless, args->debug_gpu);
    buttons_ctor(&self->buttons, &self->cpu, &self->ram, args->headless);
    if (!args->silent) {
        apu_ctor(&self->apu, &self->cpu, &self->ram, args->debug_apu);
    }
    clock_ctor(&self->clock, &self->buttons, args->frames, args->profile, args->turbo);
}

void gameboy_dtor(struct GameBoy *self) {
    gpu_dtor(&self->gpu);
    apu_dtor(&self->apu);
}

static inline void gameboy_tick(struct GameBoy *self) {
    cpu_tick(&self->cpu);
    gpu_tick(&self->gpu);
    buttons_tick(&self->buttons);
    clock_tick(&self->clock);
}

/**
 * GB CPU runs at 4MHz, but each action takes a multiple of 4 hardware
 * cycles. So to avoid overhead, we run the main loop at 1MHz, and each
 * "cycle" that each subsystem counts represents 4 hardware cycles.
 */
void gameboy_run(struct GameBoy *self) {
    while (true) {
        gameboy_tick(self);
    }
}
